pkg/util/cluster/cluster.go

```diff
@@ -346,20 +357,44 @@ func (c *Cluster) generateSubnets() (vnetPrefix string, masterSubnet string, wor
 	// 10.0.0.0/16 is used by dev-vnet to host CI
 	// 10.1.0.0/24 is used by rp-vnet to host Proxy VM
 	// 10.2.0.0/24 is used by dev-vpn-vnet to host VirtualNetworkGateway
-	var x, y int
 	rand.Seed(time.Now().UnixNano())
-	// Local Dev clusters are limited to /16 dev-vnet
-	if !c.ci {
-		x, y = 0, 2*rand.Intn(128)
-	} else {
-		x, y = rand.Intn((124))+3, 2*rand.Intn(128)
-	}
+	x, y := rand.Intn((124))+3, 2*rand.Intn(128)
 	vnetPrefix = fmt.Sprintf("10.%d.%d.0/23", x, y)
 	masterSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y)
 	workerSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y+1)
 	return
 }
 
+func (c *Cluster) getUnusedAddressSpaces(ctx context.Context, vnetResourceGroup string) (string, string, string, error) {
+	addressSpaces := make([]string, 0)
+	subnetListResultPage, err := c.subnets.List(ctx, vnetResourceGroup, devVnetName)
+	if err != nil {
+		return "", "", "", err
+	}
+	mp := make(map[string]bool)
+	for _, subnet := range subnetListResultPage.Values() {
+		mp[*subnet.AddressPrefix] = true
+	}
+
+	devVnet, err := c.vnets.Get(ctx, c.env.ResourceGroup(), devVnetName, "")
+	if err != nil {
+		return "", "", "", err
+	}
+
+	for _, addressPrefix := range *devVnet.AddressSpace.AddressPrefixes {
+		if mp[addressPrefix] {
+			continue
+		}
+		addressSpaces = append(addressSpaces, addressPrefix)
+	}
+
+	if len(addressSpaces) < 2 {
+		return "", "", "", fmt.Errorf("not enough address space available in the virtual network %s", devVnetName)
+	}
+	addressPrefix := addressSpaces[0][:len(addressSpaces[0])-1] + "3"
+	return addressPrefix, addressSpaces[0], addressSpaces[1], nil
+}
```

Comment:
I really like the idea of moving all that into its own function. Nevertheless, I think we can improve it a bit as we are defining the method on the type cluster, and it has already a lot of fields and methods. We can improve the cohesion here.

It would be a good idea to just remove the method from the struct and convert it to a standard function and pass just the interfaces and values that we need. In that way, the code is a bit better designed and easier to test as we don't need the whole cluster object or the whole c.env as we just pass from the outside what we need (the interfaces and c.env.ResourceGroup()).  Also, adding the name of the values returned in the function enhances readability as we have more than one return type of the same type (string, string, string).

I had in mind something like:
```go
func getUnusedAddressSpaces(ctx context.Context, subnets network.SubnetsClient, vnets network.VirtualNetworksClient, vnetResourceGroup string, envResourceGroup string) (addressPrefix string, masterSubnet string, workerSubnet string, err error) {
	var addressSpaces []string
	subnetListResultPage, err := subnets.List(ctx, vnetResourceGroup, devVnetName)
	if err != nil {
		return "", "", "", err
	}
	mp := make(map[string]bool)
	for _, subnet := range subnetListResultPage.Values() {
		mp[*subnet.AddressPrefix] = true
	}

	devVnet, err := vnets.Get(ctx, envResourceGroup, devVnetName, "")
	if err != nil {
		return "", "", "", err
	}

	for _, addressPrefix := range *devVnet.AddressSpace.AddressPrefixes {
		if mp[addressPrefix] {
			continue
		}
		addressSpaces = append(addressSpaces, addressPrefix)
	}

	if len(addressSpaces) < 2 {
		return "", "", "", fmt.Errorf("not enough address space available in the virtual network %s", devVnetName)
	}
	addressPrefix = addressSpaces[0][:len(addressSpaces[0])-1] + "3"
	return addressPrefix, addressSpaces[0], addressSpaces[1], nil
}
```
Also, if you want, you can use 
```go
var addressSpaces []string
```
instead of 
```go
addressSpaces := make([]string, 0)
```
as you save one memory allocation and is more idiomatic in Go.